### PR TITLE
add const assertions in typescript 3.4

### PR DESCRIPTION
**Pull request by MSakamaki**
#5 

added const assertions and read-only Array to [Prefer immutability](https://github.com/labs42io/clean-code-typescript#prefer-immutability) section.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1199,6 +1199,49 @@ interface Config {
 }
 ```
 
+case of Array, you can create a read-only array by using `ReadonlyArray<T>`.
```
Comment (dimadeveatii) [inline README.md]: Please `Capitalize` words at the beginning of a sentence.

File: README.md [hunk 2]
```diff
@@ -1199,6 +1199,49 @@ interface Config {
...
+
+declaring read-only arguments in [TypeScript 3.4 is a bit easier](https://github.com/microsoft/TypeScript/wiki/What's-new-in-TypeScript#improvements-for-readonlyarray-and-readonly-tuples).
+
+```ts
+function hoge(args: readonly string[]) {
+  args.push(1); // error
+}
+```
+
+TypeScript 3.4 introduces a new construct for literal values called [const assertions](https://github.com/microsoft/TypeScript/wiki/What's-new-in-TypeScript#const-assertions).
```
Comment (dimadeveatii) [inline README.md]: I'd suggest that you rephrase it to something like: `Prefer [const assertions](https://github.com/microsoft/TypeScript/wiki/What's-new-in-TypeScript#const-assertions) for literal values.` . No need to mention TS version, that can be checked in the link. This guide should focus on principles and we should not make it a TypeScript tutorial.

File: README.md [hunk 3]
```diff
@@ -1199,6 +1199,49 @@ interface Config {
...
+
+- no literal types in that expression should be widened (e.g. no going from "hello" to string)
```
Comment (dimadeveatii) [inline README.md]: As mentioned above, the link is included, so skip the below 3 lines.

File: README.md [hunk 4]
```diff
@@ -1199,6 +1199,49 @@ interface Config {
...
+- object literals get readonly properties
+- array literals become readonly tuples
+
+```ts
+// can not assign except the "word" string
+let hello = 'world' as const;
```
Comment (dimadeveatii) [inline README.md]: Not a good example, I'd rather use then `const hello = "hello";`

File: README.md [hunk 5]
```diff
@@ -1199,6 +1199,49 @@ interface Config {
...
+hello = 'world'; // success
+hello = 'world'; // error
+
+// read-only object
+const config = {
```
Comment (dimadeveatii) [inline README.md:1255]: This is a very good example!

**Review by dimadeveatii — CHANGES_REQUESTED** [review]
Please try to maintain the same structure, and include `**Bad:**` and `**Good:**` example sections.

File: README.md (hunk 4 above)
Comment (MSakamaki) [reply README.md]: ok, I remove this example as it is common cases for constant string.

**Comment by MSakamaki** [comment]
thanks you for the review!
I fixin it for a while to send new PR!

File: README.md [hunk 6]
```diff
@@ -1199,6 +1199,77 @@ interface Config {
 }
 ```
 
+Case of Array, you can create a read-only array by using `ReadonlyArray<T>`.
+do not allow changes such as `push()` and `fill()`, but can use features such as `concat()` and `slice()` that do not change the value.
+
+**Bad:**
+
+```ts
+const array: Array<number> = [ 1, 3, 5 ];
```
Comment (dimadeveatii) [inline README.md]: Please prefer `number[]` instead of `Array<number>`.

File: README.md [hunk 7]
```diff
@@ -1199,6 +1199,77 @@ interface Config {
...
+
+Declaring read-only arguments in [TypeScript 3.4 is a bit easier](https://github.com/microsoft/TypeScript/wiki/What's-new-in-TypeScript#improvements-for-readonlyarray-and-readonly-tuples).
+
+```ts
+function hoge(args: readonly string[]) {
+  args.push(1); // error
+}
+```
+
+Prefer [const assertions](https://github.com/microsoft/TypeScript/wiki/What's-new-in-TypeScript#const-assertions) for literal values
```
Comment (dimadeveatii) [inline README.md]: Please end the sentence with a period `.` .

File: README.md [hunk 8]
```diff
@@ -1199,6 +1199,77 @@ interface Config {
...
+
+**Bad:**
+
+```ts
+// const object
```
Comment (dimadeveatii) [inline README.md]: The comment is redundant. Declaration of `const config = { ` already says it. I'd suggest that you remove the comment.

File: README.md [hunk 9]
```diff
@@ -1199,6 +1199,77 @@ interface Config {
...
+const config = {
+  hello: 'world'
+};
+config.hello = 'world'; // value is written
```
Comment (dimadeveatii) [inline README.md]: `// value is changed`

File: README.md [hunk 10]
```diff
@@ -1199,6 +1199,77 @@ interface Config {
...
+
+// const array
+const array  = [ 1, 3, 5 ];
+array[0] = 10; // value is written
+
+// writable objects is returned
+function readonlyData(value: number) {
+  return { value };
+}
+const result = readonlyData(100);
```
Comment (dimadeveatii) [inline README.md:1247]: Please add an empty line to separate function.

**Review by dimadeveatii — CHANGES_REQUESTED** [review]
Very well. 
Please fix also the small changes I requested.

**Comment by MSakamaki** [comment]
Thank you very much for the review, I fixed the reviews issues.

**Review by dimadeveatii — APPROVED** [review]

**Comment by dimadeveatii** [comment]
Fixed #5: Add const contexts for literal expressions.
